parsers/CY.py

```diff
@@ -12,64 +12,102 @@
-    assert columns == ['Συνολική Προβλεπόμενη Ζήτηση', 'Αιολική Παραγωγή',
-        'Εκτίμηση Διεσπαρμένης Παραγωγής (Φωτοβολταϊκά και Βιομάζα)',
-        'Συνολική Ζήτηση', 'Συμβατική Παραγωγή'], 'Source format changed'
-
-    times = (m.group(1, 2, 3) for m in re.finditer(
-        '''var dateStr = "([^"]*)";\s+var hourStr = "(\d+)";\s+var minutesStr = "(\d+)";''', html))
-    prods = (m.group(1, 2, 3, 4) for m in re.finditer(
-        '''\[\s*dateStrFormat,\s*(\d+|null),\s*(\d+|null),\s*(\d+|null),\s*(\d+|null)\s*\]''', html))
-
-    data = []
-    for t, p in zip(times, prods):
-        if any(e == 'null' for e in p):
-            break
-        append_datum(data, t, p)
-
-    return data
+EXPECTED_COLUMNS = ['Συνολική Προβλεπόμενη Ζήτηση', 'Αιολική Παραγωγή',
+    'Εκτίμηση Διεσπαρμένης Παραγωγής (Φωτοβολταϊκά και Βιομάζα)',
+    'Συνολική Ζήτηση', 'Συμβατική Παραγωγή']
+
+COLUMNS_REGEX = re.compile(r'data\.addColumn\("number", "([^"]*)"\);')
+TIMES_REGEX = re.compile(
+    r'var dateStr = "([^"]*)";\s+var hourStr = "(\d+)";\s+var minutesStr = "(\d+)";')
+PRODUCTIONS_REGEX = re.compile(
+    r'\[dateStrFormat,\s*(\d+|null),\s*(\d+|null),\s*(\d+|null),\s*(\d+|null)\]')
+
+class CyprusParser:
+    session = None
+    logger: logging.Logger = None
+
+    def __init__(self, session, logger: logging.Logger):
+        self.session = session
+        self.logger = logger
+
+
+    def append_datum(self, data: list, time_str: str, prods_list: list) -> None:
+        time_value = arrow.get(time_str).replace(tzinfo='Asia/Nicosia').datetime
+
+        prods_value = {
+            'oil': prods_list[3],
+            'solar': prods_list[1],
+            'wind': prods_list[0],
+            'biomass': 0.0
+        }
+
+        # Because solar is explicitly listed as "Solar PV" (so no thermal with energy storage) and there
+        # is zero sunlight in the middle of the night (https://www.timeanddate.com/sun/cyprus/nicosia),
+        # we use the biomass+solar generation reported at 0:00 to determine the portion of biomass+solar
+        # which constitutes biomass
+        if len(data) == 0:
+            prods_value['biomass'] = prods_value['solar']
+            prods_value['solar'] = 0.0
+        else:
+            nocturnal_biomass_generation = data[0]['production']['biomass']
+            prods_value['solar'] -= nocturnal_biomass_generation
+            prods_value['biomass'] += nocturnal_biomass_generation
+        if prods_value['solar'] < 0.0:
+            # if there is (next to) no sunlight and biomass is lower than at midnight
+            prods_value['solar'] = 0.0
+
+        data.append({
+            'zoneKey': 'CY',
+            'production': prods_value,
+            'storage': {},
+            'source': 'tsoc.org.cy',
+            'datetime': time_value
+        })
+
+
+    def parse_html(self, html: str) -> list:
+        global EXPECTED_COLUMNS, COLUMNS_REGEX, TIMES_REGEX, PRODUCTIONS_REGEX
```

Comment:
In general it's best to avoid globals, these could either be class attributes or straight constants if the parser stays as a set of functions.